### PR TITLE
Langchain Setup with RAG, Chatbot and Summarisation Service

### PR DESCRIPTION
**Pull request by ghopk**
# Langchain Setup with RAG, Chatbot and Summarisation Service

This PR adds the above features to the front end. 

## Notes:
- A better choice of model for summarisation should be heavily considered, current model is `Falconsai/text_summarization`
- The embeddings for the vector database and the LLM model accessed by lemonfox are not matching, this may not cause an issue depending on how langchain retrieves the relevant documents to be passed to the LLM
- Speak to @ghopk to get a .env file to get the correct OpenAI API key

### CODE REVIEW

File: src/front_end/similarity_and_summarisation_service.py [hunk 1]
```diff
@@ -1,41 +1,93 @@
...
 
+load_dotenv()
+OPENAI_API_KEY = os.getenv("OPENAI_API_KEY")
+BASE_URL = os.getenv("BASE_URL")
 
-def _load_vector_database():
-    chroma_client = chromadb.PersistentClient(
-        path=VECTOR_DATABASE_PATH,
+
+def _load_langchain_chroma_vector_database():
```
Comment (michal-gnacik-ab) [inline src/front_end/similarity_and_summarisation_service.py]: return type hint -> Chroma

File: src/front_end/similarity_and_summarisation_service.py [hunk 2]
```diff
@@ -1,41 +1,93 @@
...
+from langchain_community.chat_models import ChatOpenAI
+from transformers import pipeline
+from dotenv import load_dotenv
+import os
 from src.constants.paths import VECTOR_DATABASE_PATH
 from src.constants.parameters import VECTOR_DATABASE_NAME
 
+load_dotenv()
+OPENAI_API_KEY = os.getenv("OPENAI_API_KEY")
+BASE_URL = os.getenv("BASE_URL")
```
Comment (michal-gnacik-ab) [inline src/front_end/similarity_and_summarisation_service.py]: Keep these are the bottom of the script under 

if __name__ == "__main__":

File: src/front_end/similarity_and_summarisation_service.py [hunk 3]
```diff
@@ -1,41 +1,93 @@
...
-    results = collection.query(
-        query_texts=[query_text],
-        n_results=5,
+def _abstract_summarisation(relevant_docs_df: pd.DataFrame) -> pd.DataFrame:
+    summariser = pipeline("summarization", model="Falconsai/text_summarization")
+    relevant_docs_df["Summary"] = relevant_docs_df["Abstract"].apply(summariser)
+    return relevant_docs_df
+
+
+def _create_similar_results_dataframe(results: list):
```
Comment (michal-gnacik-ab) [inline src/front_end/similarity_and_summarisation_service.py]: list[Any] import Any from collections. Return type missing.

File: src/front_end/similarity_and_summarisation_service.py [hunk 4]
```diff
@@ -1,41 +1,93 @@
...
+    return chroma_db
 
 
-def retrieve_similar_docs_and_summarisation_from_query(query_text: str):
-    collection = _load_vector_database()
-    results = collection.query(
-        query_texts=[query_text],
-        n_results=5,
+def _abstract_summarisation(relevant_docs_df: pd.DataFrame) -> pd.DataFrame:
+    summariser = pipeline("summarization", model="Falconsai/text_summarization")
```
Comment (michal-gnacik-ab) [inline src/front_end/similarity_and_summarisation_service.py]: Consider moving "Falconsai/text_summarization" to a config file an import as Global.

File: src/front_end/similarity_and_summarisation_service.py [hunk 5]
```diff
@@ -1,41 +1,93 @@
...
+    relevant_docs_df["Summary"] = relevant_docs_df["Abstract"].apply(summariser)
+    return relevant_docs_df
+
+
+def _create_similar_results_dataframe(results: list):
+    doc_titles = [results[i].metadata["title"] for i in range(len(results))]
+    doc_urls = [results[i].metadata["url"] for i in range(len(results))]
+    doc_abstracts = [results[i].metadata["abstract"] for i in range(len(results))]
```
Comment (michal-gnacik-ab) [inline src/front_end/similarity_and_summarisation_service.py]: Set len(results) as a variable, so then len(results) is not calculated three times. Change the name of the index from one letter i to index. Avoid using single letter variabels. 

Consider having a helper function 

```
def _extract_metadata(result: list[Any], metadata_type: str) -> list[Any]:
    length_results_list = len(results)
    doc_property =  [results[index].metadata[metadata_type] for index in range(length_results_list)]
    return doc_property
```

File: src/front_end/similarity_and_summarisation_service.py [hunk 6]
```diff
@@ -1,41 +1,93 @@
...
+    )
+    response = chain(query_text)["result"]
+    return response
 
-def summarise_relevant_documents(similar_docs: pd.DataFrame):
-    # LLM summarisation code here
-    summarised_text = "Summarised text"
 
-    return summarised_text
+def retrieve_similar_docs_and_summarisation_from_query(query_text: str):
```
Comment (michal-gnacik-ab) [inline src/front_end/similarity_and_summarisation_service.py]: Output hints tuple[pd.DataFrame, str, ...]

File: src/front_end/similarity_and_summarisation_service.py [hunk 7]
```diff
@@ -1,41 +1,93 @@
...
+    chroma_db = _load_langchain_chroma_vector_database()
+    results = chroma_db.similarity_search(query_text, k=5)
```
Comment (michal-gnacik-ab) [inline src/front_end/similarity_and_summarisation_service.py]: k should not be hardcoded.

File: src/llm_model/langchain_pipeline.py [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+generation_args = {
+    "max_new_tokens": 500,
+    "return_full_text": False,
+    "temperature": 0.0,
+    "do_sample": False,
+}
+
+output = pipe(messages, **generation_args)
+print(output[0]['generated_text'])
+
```
Comment (michal-gnacik-ab) [inline src/llm_model/langchain_pipeline.py:132]: The pipeline should be captured in a python function that we call under 
`if __name__ == "__main__":`

this looks very notebooky.

**Review by michal-gnacik-ab — COMMENTED** [review]
Ok left some comments, for the first review iteration. Mostly to improve code presentation.

File: src/llm_model/langchain_pipeline.py (hunk 8 above)
Comment (michal-gnacik-ab) [reply src/llm_model/langchain_pipeline.py:132]: I'm happy to elaborate on this one if unclear.

File: src/front_end/similarity_and_summarisation_service.py (hunk 5 above)
Comment (ghopk) [reply src/front_end/similarity_and_summarisation_service.py]: Helper function is good but then len(results) still gets calculated multiple times. I don't see the whole point of passing in len(reuslts) into the function either so have left it as your implementation

File: src/llm_model/langchain_pipeline.py (hunk 8 above)
Comment (ghopk) [reply src/llm_model/langchain_pipeline.py:132]: This is a databricks notebook!

**Review by michal-gnacik-ab — COMMENTED** [review]
If this is a file to run frontend, I would split it into a few files. 
One that has frontend layout, and the other ones having frontend functionality. 
Currently, `src/front_end/similarity_and_summarisation_service.py` is a bit messy as it has everything in it.

**Comment by ghopk** [comment]
> If this is a file to run frontend, I would split it into a few files. One that has frontend layout, and the other ones having frontend functionality. Currently, `src/front_end/similarity_and_summarisation_service.py` is a bit messy as it has everything in it.

Definitely agree, I thought the same, will split these out into separate files.

File: src/constants/parameters.py [hunk 9]
```diff
@@ -1 +1,4 @@
-VECTOR_DATABASE_NAME = "vector_database"
+VECTOR_DATABASE_NAME = "medical_research_papers"
```
Comment (Ali-Hindwan) [inline src/constants/parameters.py:1]: @ghopk , now we have one vector database for both tasks?

**Pull request merged**